File: Labs/lab5/Dictionary.c
//------------------------------------------------
// Wilson Au-Yeung
// wwauyeun
// May 6,2016
// CMPS12B
// Dictionary.c
// Linked List implementation of the Dictionary.ADT.
//------------------------------------------------
#include "Dictionary.h"
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// NodeObj
typedef struct NodeObj {
  struct NodeObj *next;
  char *key;
  char *value;
} NodeObj;

// Node
typedef NodeObj *Node;

// newNode()
// constructor for private Node Type
Node newNode(char *k, char *v) {
  Node N = malloc(sizeof(NodeObj));
  assert(N != NULL);
  N->key = k;
  N->value = v;
  N->next = NULL;
  return (N);
}
void freeNode(Node *pN) {
  if (pN != NULL && *pN != NULL) {
    free(*pN);
    *pN = NULL;
  }
}
struct DictionaryObj {
  Node Head;
  Node Tail;
  int numPairs;
} DictionaryObj;
// Constructor for the new Dictionary Type
Dictionary newDictionary(void) {
  Dictionary D = malloc(sizeof(Dictionary));
  assert(D != NULL);
  D->Head = NULL;
  D->Tail = NULL;
  D->numPairs = 0;
  return D;
}
// destructor for the Dictionary
void freeDictionary(Dictionary *pD) {
  if (pD != NULL && *pD != NULL) {
    makeEmpty(*pD);
    free(*pD);
    *pD = NULL;
  }
}
// it checks if the Dictionary type is empty. Returns 1 if it is, and if not
// returns 0
int isEmpty(Dictionary D) {
  if (D == NULL) {
    fprintf(
        stderr,
        "Dictionary Error: calling isEmpty() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  if (D->numPairs == 0) {
    return 1;
  } else
    return 0;
}
// the numPairs in the Dictionary Type keeps track of how many pairs are in D
int size(Dictionary D) {
  if (D == NULL) {
    fprintf(stderr,
            "Dictionary Error: calling size() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  return D->numPairs;
}
// allows me to insert a new key that goes with a value into the Dictionary.
void insert(Dictionary D, char *k, char *v) {
  if (D == NULL) {
    fprintf(
        stderr,
        "Dictionary Error: calling insert() on NULL Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  if (lookup(D, k) != NULL) {
    fprintf(stderr, "Dictionary Error: calling insert() duplicate keys\n");
    exit(EXIT_FAILURE);
  }
  if (D->numPairs == 0) {
    D->Head = D->Tail = newNode(k, v);
  } else {
    Node N = newNode(k, v);
    D->Tail->next = N;
    D->Tail = N;
  }
  D->numPairs++;
}
// this helps me check if there are duplicates of a key or if there are any of
// the keys at all used by insert() and delete()
char *lookup(Dictionary D, char *k) {
  Node N = D->Head;
  if (D == NULL) {
    fprintf(stderr, "Dictionary Error: calling lookup()  on NULL Dictionary\n");
    exit(EXIT_FAILURE);
  }
  for (N; N != NULL; N = N->next) {
    if (strcmp(N->key, k) == 0)
      return N->value;
  }
  return NULL;
}
// empties the Dictionary to NULL
void makeEmpty(Dictionary D) {
  D->Head = NULL;
  D->Tail = NULL;
  D->numPairs = 0;
}
// deletes a key and its value in a dictionary
void delete (Dictionary D, char *k) {
  Node N = D->Head;
  if (lookup(D, k) == NULL) {
    fprintf(stderr, "Dictionary error: key not found\n");
    exit(EXIT_FAILURE);
  }
  if (strcmp(N->key, k) == 0) {
    Node Temp = D->Head;
    Node non = Temp->next;
    D->Head = non;
    freeNode(&Temp);
  } else {
    while (N != NULL && N->next != NULL) {
      if (strcmp(N->next->key, k) == 0) {
        Node non = N;
        Node non2 = N->next;
        non->next = non2->next;
        N = non;
        freeNode(&non2);
      }
      N = N->next;
    }
  }
  D->numPairs--;
}

// prints out the Dictionary
void printDictionary(FILE *out, Dictionary D) {
  Node N;
  if (D == NULL) {
    fprintf(stderr, "Dictionary Error: calling printDictionary() on NULL "
                    "Dictionary reference\n");
    exit(EXIT_FAILURE);
  }
  for (N = D->Head; N != NULL; N = N->next)
    fprintf(out, "%s %s \n", N->key, N->value);
  fprintf(out, "\n");
}

File: Labs/lab7/Dictionary.java
//--------------------------------------------------------------
// Dictioanry.java
// Wilson Au-Yeung
// wwauyeun
// May 31,2016
// CMPS12B
// This file is a Binary Search Tree implementation of the Dictionary ADT
//--------------------------------------------------------------
public class Dictionary implements DictionaryInterface {
  // Private Node Class
  private class Node {
    Pair item;
    Node right;
    Node left;
    Node(Pair x) {
      item = x;
      // this.value = value;
      right = null;
      left = null;
    }
  }
  // private Pair Class
  private class Pair {
    String key;
    String value;
    Pair(String a, String b) {
      key = a;
      value = b;
    }
  }

  private Node root;
  private int numPairs;
  // Dictionary constructor
  public Dictionary() {
    root = null;
    numPairs = 0;
  }

  public boolean isEmpty() {
    if (numPairs == 0)
      return true;
    else
      return false;
  }
  public int size() { return numPairs; }
  // returns the parent of N
  private Node findParent(Node N, Node R) {
    Node P = null;
    if (N != R) {
      P = R;
      while (P.left != N && P.right != N) {
        if (N.item.key.compareToIgnoreCase(P.item.key) < 0) {
          P = P.left;
        } else
          P = P.right;
      }
    }
    return P;
  }
  // finds and returns the Node that has a key searched for
  private Node findKey(Node R, String k) {
    if (R == null || R.item.key.equals(k))
      return R;
    else if (R.item.key.compareToIgnoreCase(k) > 0)
      return findKey(R.left, k);
    else
      return findKey(R.right, k);
  }
  // returns the leftmost Node in the subtree
  private Node findLeftmost(Node R) {
    Node L = R;
    for (L = R; L.left != null; L = L.left) {
    }
    return L;
  }
  // inserts a pair into the dictionary adt
  public void insert(String k, String v) throws DuplicateKeyException {
    Node N, A, B;
    if (findKey(root, k) != null) {
      throw new DuplicateKeyException("Cannot insert duplicate key");
    }
    N = new Node(new Pair(k, v));
    B = null;
    A = root;
    while (A != null) {
      B = A;
      if (A.item.key.compareToIgnoreCase(k) > 0)
        A = A.left;
      else
        A = A.right;
    }
    if (B == null)
      root = N;
    else if (B.item.key.compareToIgnoreCase(k) > 0)
      B.left = N;
    else
      B.right = N;
    numPairs++;
  }

  // removes a pair from the dictionary ADT
  public void delete(String key)throws KeyNotFoundException {
    Node N, P, S;
    if (findKey(root, key) == null) {
      throw new KeyNotFoundException(
          "Dictionary Error: delete() cannot delete non-existent key");
    }
    N = findKey(root, key);
    if (N.left == null && N.right == null) {
      if (N == root)
        root = null;
      else {
        P = findParent(N, root);
        if (P.right == N)
          P.right = null;
        else
          P.left = null;
      }
    } else if (N.right == null) {
      if (N == root)
        root = N.left;
      else {
        P = findParent(N, root);
        if (P.right == N)
          P.right = N.left;
        else
          P.left = N.left;
      }
    } else if (N.left == null) {
      if (N == root)
        root = N.right;

      else {
        P = findParent(N, root);
        if (P.right == N)
          P.right = N.right;
        else
          P.left = N.right;
      }
    } else {
      S = findLeftmost(N.right);
      N.item.key = S.item.key;
      N.item.value = S.item.value;
      P = findParent(S, N);
      if (P.right == S)
        P.right = S.right;
      else
        P.left = S.right;
    }
    numPairs--;
  }
  public String lookup(String key) {
    Node N;
    N = findKey(root, key);
    return (N == null ? null : N.item.value);
  }
  public void makeEmpty() {
    deleteAll(root);
    root = null;
    numPairs = 0;
  }
  public String toString() {
    printInOrder(root);
    return "";
  }
  void printInOrder(Node R) {
    if (R != null) {
      printInOrder(R.left);
      System.out.println(R.item.key + " " + R.item.value);
      printInOrder(R.right);
    }
  }
  // deletes all the branches and nodes in the tree
  void deleteAll(Node N) {
    if (N != null) {
      deleteAll(N.left);
      deleteAll(N.right);
    }
  }
}
